tensorflow/stream_executor/rocm/rocm_blas.cc

```diff
@@ -1969,27 +2121,120 @@ port::Status ROCMBlas::DoBlasGemmBatchedInternal(
+  }
+
+  if (ROCMBlasTranspose(transb) == rocblas_operation_none) {
+    assert(ldb >= k);
+    batch_stride_b = ldb * n;
   } else {
-    return port::Status(port::error::INTERNAL,
-                        "failed BLAS call, see log for details");
+    assert(ldb >= n);
+    batch_stride_b = ldb * k;
+  }
+
+  // Alocate local vectors to hold device pointers to matrices
+  std::vector<MAPPED_T *> a_raw_ptrs, b_raw_ptrs, c_raw_ptrs;
+  for (int i = 0; i < batch_count; ++i) {
+    // static_cast does work when converting Eigen::half* to rocblas_half*,
+    // hence the use of reinterpret_cast
+    a_raw_ptrs.push_back(
+        reinterpret_cast<MAPPED_T *>(a_ptrs_to_wrappers[i]->opaque()));
+    b_raw_ptrs.push_back(
+        reinterpret_cast<MAPPED_T *>(b_ptrs_to_wrappers[i]->opaque()));
+    c_raw_ptrs.push_back(
+        reinterpret_cast<MAPPED_T *>(c_ptrs_to_wrappers[i]->opaque()));
+  }
+
+  DeviceMemory<MAPPED_T> a;
+  // Make sure the temporary memory are in-scope before the function returns
+  std::unique_ptr<TemporaryDeviceMemory<MAPPED_T>> a_temp;
+  bool reallocated_a, reallocated_b, reallocated_c;
+  port::Status a_allocation_status =
+      AllocateStridedBuffer<T>(a_raw_ptrs, batch_count, batch_stride_a*2,
+                               scratch_allocator, stream, &a_temp, &a, true, reallocated_a);
+  if (a_allocation_status != port::Status::OK()) {
+    return a_allocation_status;
+  }
+
+  DeviceMemory<MAPPED_T> b;
+  std::unique_ptr<TemporaryDeviceMemory<MAPPED_T>> b_temp;
+  port::Status b_allocation_status =
+      AllocateStridedBuffer<T>(b_raw_ptrs, batch_count, batch_stride_b*2,
+                               scratch_allocator, stream, &b_temp, &b, true, reallocated_b);
+  if (b_allocation_status != port::Status::OK()) {
+    return b_allocation_status;
   }
+
+  DeviceMemory<MAPPED_T> c;
+  std::unique_ptr<TemporaryDeviceMemory<MAPPED_T>> c_temp;
+  port::Status c_allocation_status =
+      AllocateStridedBuffer<T>(c_raw_ptrs, batch_count, batch_stride_c*2,
+                               scratch_allocator, stream, &c_temp, &c, true, reallocated_c); // can disable copy if beta=0
+  if (c_allocation_status != port::Status::OK()) {
+    return c_allocation_status;
+  }
+  hipDeviceSynchronize();
+
+  auto status = DoBlasComplex(scratch_allocator, stream,
+    (const T*)a.opaque(), (const T*)b.opaque(), (T*)c.opaque(),
+    batch_stride_a*batch_count, batch_stride_b*batch_count, batch_stride_c*batch_count,
+    alpha, beta,
+    rocblas_func,
+    ((transa == blas::Transpose::kConjugateTranspose) ? 1 : 0) + ((transb == blas::Transpose::kConjugateTranspose) ? 2 : 0),
+    ROCMBlasTranspose(transa), ROCMBlasTranspose(transb),
+    batch_stride_a, batch_stride_b, batch_stride_c, batch_count,
+    m, n, k, lda, ldb, ldc);
+  if(status!=port::Status::OK())
+    return status;
+  if(reallocated_c)
+      return ReorganizeMemory(stream, &c, c_raw_ptrs, batch_count, batch_stride_c*2, false);
+  return port::Status::OK();
```

Comment:
If I understand right, the only difference between this function (`DoBlasGemmBatchedInternalCpx`) and the `DoBlasGemmBatchedInternal` is here, `DoBlasComplex()` is invoked instead of `DoBlasInternal`. Could you consolidate the two into one to avoid duplicated code in between?